File: src/security/security-handler.ts
/**
 * Main security handler that ties together encryption/decryption for PDF documents.
 */

import type { PdfObject, PdfRef, PdfDict } from '../core/types.js';
import { dictGetNumber, dictGetName, dictGetBool, dictGetDict } from '../core/objects.js';
import type { EncryptionOptions } from './encryption-dict.js';
import { createEncryptionDict } from './encryption-dict.js';
import { encryptObject } from './encrypt-stream.js';
import { decryptObject } from './decrypt-stream.js';
import {
  padPassword,
  PDF_PASSWORD_PADDING,
  computeOwnerKey,
  computeUserKey,
  validateUserPasswordR6,
  validateOwnerPasswordR6,
  recoverFileKeyFromUserR6,
  recoverFileKeyFromOwnerR6,
} from './password.js';
import { md5 } from './md5.js';
import { rc4 } from './rc4.js';

/**
 * SecurityHandler manages encryption and decryption for a PDF document.
 */
export class SecurityHandler {
  private encryptionKey: Uint8Array;
  private algorithm: string;
  private revision: number;

  private constructor(encryptionKey: Uint8Array, algorithm: string, revision: number) {
    this.encryptionKey = encryptionKey;
    this.algorithm = algorithm;
    this.revision = revision;
  }

  /**
   * Create a new security handler for encrypting a PDF document.
   *
   * @param options - Encryption options (algorithm, passwords, permissions)
   * @param fileId - The first element of the /ID array
   * @returns The security handler and the /Encrypt dictionary
   */
  static create(
    options: EncryptionOptions,
    fileId: Uint8Array,
  ): { handler: SecurityHandler; encryptDict: PdfDict } {
    const { encryptDict, encryptionKey } = createEncryptionDict(options, fileId);

    let revision: number;
    switch (options.algorithm) {
      case 'rc4-40': revision = 2; break;
      case 'rc4-128': revision = 3; break;
      case 'aes-128': revision = 4; break;
      case 'aes-256': revision = 6; break;
      default: throw new Error(`Unknown algorithm: ${options.algorithm}`);
    }

    const handler = new SecurityHandler(encryptionKey, options.algorithm, revision);
    return { handler, encryptDict };
  }

  /**
   * Create a security handler from an existing /Encrypt dictionary (for decryption).
   * Validates the password and derives the encryption key.
   *
   * @param encryptDict - The parsed /Encrypt dictionary
   * @param password - The password to try (tested as user password first, then owner)
   * @param fileId - The first element of the /ID array
   * @returns The security handler ready for decryption
   * @throws Error if the password is invalid
   */
  static fromEncryptDict(
    encryptDict: PdfDict,
    password: string,
    fileId: Uint8Array,
  ): SecurityHandler {
    const V = dictGetNumber(encryptDict, 'V') ?? 0;
    const R = dictGetNumber(encryptDict, 'R') ?? 0;
    const lengthBits = dictGetNumber(encryptDict, 'Length') ?? 40;
    const P = dictGetNumber(encryptDict, 'P') ?? 0;
    const encryptMetadata = dictGetBool(encryptDict, 'EncryptMetadata') ?? true;

    // Extract /O and /U entries as raw bytes
    const oEntry = extractStringBytes(encryptDict, 'O');
    const uEntry = extractStringBytes(encryptDict, 'U');

    if (!oEntry || !uEntry) {
      throw new Error('Missing /O or /U entry in encryption dictionary');
    }

    // Determine algorithm from V and R
    let algorithm: string;
    if (R === 6) {
      algorithm = 'aes-256';
    } else if (R === 4) {
      // Check CF for AESV2
      const cf = dictGetDict(encryptDict, 'CF');
      if (cf) {
        const stdCF = dictGetDict(cf, 'StdCF');
        if (stdCF) {
          const cfm = dictGetName(stdCF, 'CFM');
          algorithm = cfm === 'AESV2' ? 'aes-128' : 'rc4-128';
        } else {
          algorithm = 'rc4-128';
        }
      } else {
        algorithm = 'rc4-128';
      }
    } else if (V === 2) {
      algorithm = 'rc4-128';
    } else {
      algorithm = 'rc4-40';
    }

    // Rev 6: AES-256
    if (R === 6) {
      return SecurityHandler.fromEncryptDictR6(encryptDict, password, uEntry, oEntry);
    }

    // Rev 2-4: try as user password first
    const keyLength = lengthBits;
    const { userEntry: computedU, encryptionKey } = computeUserKey(
      password, oEntry, P, fileId, R, keyLength, encryptMetadata,
    );

    // Validate user password
    if (validateUserPassword(computedU, uEntry, R)) {
      return new SecurityHandler(encryptionKey, algorithm, R);
    }

    // Try as owner password
    const ownerKey = deriveOwnerDecryptionKey(password, oEntry, R, keyLength);
    const userPassword = recoverUserPasswordFromOwner(ownerKey, oEntry, R, keyLength);

    const { userEntry: computedU2, encryptionKey: encKey2 } = computeUserKey(
      userPassword, oEntry, P, fileId, R, keyLength, encryptMetadata,
    );

    if (validateUserPassword(computedU2, uEntry, R)) {
      return new SecurityHandler(encKey2, algorithm, R);
    }

    throw new Error('Invalid password');
  }

  /**
   * Handle Rev 6 (AES-256) password validation and key recovery.
   */
  private static fromEncryptDictR6(
    encryptDict: PdfDict,
    password: string,
    uEntry: Uint8Array,
    oEntry: Uint8Array,
  ): SecurityHandler {
    const ueEntry = extractStringBytes(encryptDict, 'UE');
    const oeEntry = extractStringBytes(encryptDict, 'OE');

    if (!ueEntry || !oeEntry) {
      throw new Error('Missing /UE or /OE entry for Rev 6 encryption');
    }

    // Try as user password
    if (validateUserPasswordR6(password, uEntry)) {
      const encryptionKey = recoverFileKeyFromUserR6(password, uEntry, ueEntry);
      return new SecurityHandler(encryptionKey, 'aes-256', 6);
    }

    // Try as owner password
    if (validateOwnerPasswordR6(password, oEntry, uEntry)) {
      const encryptionKey = recoverFileKeyFromOwnerR6(password, oEntry, uEntry, oeEntry);
      return new SecurityHandler(encryptionKey, 'aes-256', 6);
    }

    throw new Error('Invalid password');
  }

  /**
   * Encrypt a PDF object.
   */
  encryptObject(obj: PdfObject, ref: PdfRef): PdfObject {
    return encryptObject(obj, ref, this.encryptionKey, this.algorithm, this.revision);
  }

  /**
   * Decrypt a PDF object.
   */
  decryptObject(obj: PdfObject, ref: PdfRef): PdfObject {
    return decryptObject(obj, ref, this.encryptionKey, this.algorithm, this.revision);
  }

  getEncryptionKey(): Uint8Array {
    return this.encryptionKey;
  }

  getAlgorithm(): string {
    return this.algorithm;
  }

  getRevision(): number {
    return this.revision;
  }
}

/**
 * Extract raw bytes from a string entry in a dict.
 */
function extractStringBytes(dict: PdfDict, key: string): Uint8Array | undefined {
  const obj = dict.entries.get(key);
  if (obj && obj.type === 'string') {
    return obj.value;
  }
  return undefined;
}

/**
 * Validate user password by comparing /U entries.
 * Rev 2: compare all 32 bytes.
 * Rev 3-4: compare first 16 bytes only (rest is arbitrary padding).
 */
function validateUserPassword(
  computedU: Uint8Array,
  storedU: Uint8Array,
  revision: number,
): boolean {
  const compareLen = revision === 2 ? 32 : 16;
  if (computedU.length < compareLen || storedU.length < compareLen) return false;
  for (let i = 0; i < compareLen; i++) {
    if (computedU[i] !== storedU[i]) return false;
  }
  return true;
}

/**
 * Derive the owner key for decrypting the /O value to recover the user password.
 */
function deriveOwnerDecryptionKey(
  ownerPassword: string,
  _ownerEntry: Uint8Array,
  revision: number,
  keyLength: number,
): Uint8Array {
  const keyBytes = keyLength / 8;
  const paddedOwner = padPassword(ownerPassword);
  let hash = md5(paddedOwner);

  if (revision >= 3) {
    for (let i = 0; i < 50; i++) {
      hash = md5(hash);
    }
  }

  return hash.subarray(0, keyBytes);
}

/**
 * Recover the user password from the /O entry using the owner key.
 * This is the reverse of the owner key computation (Algorithm 7 from PDF spec).
 */
function recoverUserPasswordFromOwner(
  ownerKey: Uint8Array,
  ownerEntry: Uint8Array,
  revision: number,
  keyLength: number,
): string {
  const keyBytes = keyLength / 8;
  let decrypted = ownerEntry;

  if (revision === 2) {
    decrypted = rc4(ownerKey, ownerEntry);
  } else {
    // Rev 3+: decrypt with keys from 19 down to 0
    decrypted = new Uint8Array(ownerEntry);
    for (let i = 19; i >= 0; i--) {
      const modKey = new Uint8Array(keyBytes);
      for (let j = 0; j < keyBytes; j++) {
        modKey[j] = ownerKey[j] ^ i;
      }
      decrypted = rc4(modKey, decrypted);
    }
  }

  // The decrypted value is the padded user password
  // Convert back to string (take bytes before padding starts)
  let len = 32;
  // Find where the padding constant starts
  for (let i = 0; i < 32; i++) {
    if (decrypted[i] === PDF_PASSWORD_PADDING[0]) {
      // Check if the rest matches the padding
      let match = true;
      for (let j = 0; j + i < 32 && j < 32; j++) {
        if (decrypted[i + j] !== PDF_PASSWORD_PADDING[j]) {
          match = false;
          break;
        }
      }
      if (match) {
        len = i;
        break;
      }
    }
  }

  let result = '';
  for (let i = 0; i < len; i++) {
    result += String.fromCharCode(decrypted[i]);
  }
  return result;
}

File: src/security/security_handler.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const types = @import("../core/types.zig");
const PdfObject = types.PdfObject;
const md5_mod = @import("md5.zig");
const rc4_mod = @import("rc4.zig");
const aes_mod = @import("aes.zig");

/// Encryption algorithm selection.
pub const Algorithm = enum {
    rc4_40,
    rc4_128,
    aes_128,
    aes_256,

    /// Returns the key length in bytes for this algorithm.
    pub fn keyLength(self: Algorithm) usize {
        return switch (self) {
            .rc4_40 => 5,
            .rc4_128 => 16,
            .aes_128 => 16,
            .aes_256 => 32,
        };
    }

    /// Returns the PDF /V value for this algorithm.
    pub fn vValue(self: Algorithm) i64 {
        return switch (self) {
            .rc4_40 => 1,
            .rc4_128 => 2,
            .aes_128 => 4,
            .aes_256 => 5,
        };
    }

    /// Returns the PDF /R value (revision) for this algorithm.
    pub fn rValue(self: Algorithm) i64 {
        return switch (self) {
            .rc4_40 => 2,
            .rc4_128 => 3,
            .aes_128 => 4,
            .aes_256 => 6,
        };
    }
};

/// PDF document permissions (maps to the /P value in the encrypt dict).
pub const Permissions = packed struct {
    printing: bool = true,
    modifying: bool = true,
    copying: bool = true,
    annotating: bool = true,
    filling_forms: bool = true,
    content_accessibility: bool = true,
    document_assembly: bool = true,
    high_quality_printing: bool = true,

    /// Encode permissions into the 32-bit /P value per PDF spec.
    pub fn toInt(self: Permissions) i32 {
        // Bits 1-2 must be 0, bits 7-8 reserved (must be 1 for rev 2).
        // Bit 3: print, Bit 4: modify, Bit 5: copy, Bit 6: annotate
        // Bit 9: fill forms, Bit 10: extract for accessibility
        // Bit 11: assemble, Bit 12: high-quality print
        var p: i32 = -3904; // All upper bits set, bits 1-2 clear, bits 7-8 set
        if (self.printing) p |= (1 << 2);
        if (self.modifying) p |= (1 << 3);
        if (self.copying) p |= (1 << 4);
        if (self.annotating) p |= (1 << 5);
        if (self.filling_forms) p |= (1 << 8);
        if (self.content_accessibility) p |= (1 << 9);
        if (self.document_assembly) p |= (1 << 10);
        if (self.high_quality_printing) p |= (1 << 11);
        return p;
    }
};

/// Options for initializing encryption.
pub const EncryptionOptions = struct {
    owner_password: []const u8 = "",
    user_password: []const u8 = "",
    algorithm: Algorithm = .aes_128,
    permissions: Permissions = .{},
};

/// PDF padding string (32 bytes, per PDF spec).
const pdf_padding = [_]u8{
    0x28, 0xBF, 0x4E, 0x5E, 0x4E, 0x75, 0x8A, 0x41,
    0x64, 0x00, 0x4E, 0x56, 0xFF, 0xFA, 0x01, 0x08,
    0x2E, 0x2E, 0x00, 0xB6, 0xD0, 0x68, 0x3E, 0x80,
    0x2F, 0x0C, 0xA9, 0xFE, 0x64, 0x53, 0x69, 0x7A,
};

/// Handles PDF document encryption: key computation, stream encryption,
/// and building the /Encrypt dictionary.
pub const SecurityHandler = struct {
    algorithm: Algorithm,
    owner_password_hash: [32]u8,
    user_password_hash: [32]u8,
    encryption_key: [32]u8,
    key_length: usize,
    permissions: Permissions,

    /// Initialize a security handler from the given options.
    /// Computes owner/user password hashes and the encryption key.
    pub fn init(options: EncryptionOptions) SecurityHandler {
        var handler = SecurityHandler{
            .algorithm = options.algorithm,
            .owner_password_hash = undefined,
            .user_password_hash = undefined,
            .encryption_key = undefined,
            .key_length = options.algorithm.keyLength(),
            .permissions = options.permissions,
        };

        handler.computeOwnerPasswordHash(options.owner_password, options.user_password);
        handler.computeEncryptionKey(options.user_password);
        handler.computeUserPasswordHash();

        return handler;
    }

    /// Pad or truncate a password to exactly 32 bytes using the PDF padding string.
    fn padPassword(password: []const u8) [32]u8 {
        var result: [32]u8 = pdf_padding;
        const copy_len = @min(password.len, 32);
        @memcpy(result[0..copy_len], password[0..copy_len]);
        return result;
    }

    fn computeOwnerPasswordHash(self: *SecurityHandler, owner_pwd: []const u8, user_pwd: []const u8) void {
        // Use owner password if provided, otherwise use user password
        const pwd = if (owner_pwd.len > 0) owner_pwd else user_pwd;
        const padded = padPassword(pwd);

        // MD5 hash of the padded owner password
        var hash = md5_mod.md5(&padded);

        // For revision 3+, hash 50 more times
        if (self.algorithm.rValue() >= 3) {
            for (0..50) |_| {
                hash = md5_mod.md5(hash[0..self.key_length]);
            }
        }

        // Use the hash as a key to RC4-encrypt the padded user password
        const user_padded = padPassword(user_pwd);
        var encrypted: [32]u8 = undefined;
        rc4_mod.rc4(hash[0..self.key_length], &user_padded, &encrypted);

        // For revision 3+, iterate 19 times with modified keys
        if (self.algorithm.rValue() >= 3) {
            for (1..20) |i| {
                var modified_key: [32]u8 = undefined;
                for (0..self.key_length) |k| {
                    modified_key[k] = hash[k] ^ @as(u8, @intCast(i));
                }
                var temp: [32]u8 = undefined;
                rc4_mod.rc4(modified_key[0..self.key_length], &encrypted, &temp);
                encrypted = temp;
            }
        }

        self.owner_password_hash = encrypted;
    }

    fn computeEncryptionKey(self: *SecurityHandler, user_pwd: []const u8) void {
        const padded = padPassword(user_pwd);

        // Build the hash input: padded password + O value + P value + file ID
        // For simplicity, use a fixed file ID (in production, this would be the document's /ID)
        var hasher = std.crypto.hash.Md5.init(.{});
        hasher.update(&padded);
        hasher.update(&self.owner_password_hash);

        // Permissions as little-endian 4 bytes
        const p = self.permissions.toInt();
        const p_bytes = std.mem.toBytes(p);
        hasher.update(&p_bytes);

        // File ID (using a fixed placeholder)
        const file_id = [_]u8{0} ** 16;
        hasher.update(&file_id);

        var hash: [16]u8 = undefined;
        hasher.final(&hash);

        // For revision 3+, hash 50 more times
        if (self.algorithm.rValue() >= 3) {
            for (0..50) |_| {
                const temp = md5_mod.md5(hash[0..self.key_length]);
                @memcpy(hash[0..16], &temp);
            }
        }

        @memcpy(self.encryption_key[0..self.key_length], hash[0..self.key_length]);
    }

    fn computeUserPasswordHash(self: *SecurityHandler) void {
        if (self.algorithm.rValue() >= 3) {
            // Hash the padding string with file ID
            var hasher = std.crypto.hash.Md5.init(.{});
            hasher.update(&pdf_padding);
            const file_id = [_]u8{0} ** 16;
            hasher.update(&file_id);
            var hash: [16]u8 = undefined;
            hasher.final(&hash);

            // RC4 encrypt with the encryption key
            var encrypted: [16]u8 = undefined;
            rc4_mod.rc4(self.encryption_key[0..self.key_length], &hash, &encrypted);

            // 19 iterations with modified keys
            for (1..20) |i| {
                var modified_key: [32]u8 = undefined;
                for (0..self.key_length) |k| {
                    modified_key[k] = self.encryption_key[k] ^ @as(u8, @intCast(i));
                }
                var temp: [16]u8 = undefined;
                rc4_mod.rc4(modified_key[0..self.key_length], &encrypted, &temp);
                encrypted = temp;
            }

            // Pad to 32 bytes
            @memcpy(self.user_password_hash[0..16], &encrypted);
            @memset(self.user_password_hash[16..32], 0);
        } else {
            // Revision 2: RC4 encrypt the padding with the key
            rc4_mod.rc4(self.encryption_key[0..self.key_length], &pdf_padding, &self.user_password_hash);
        }
    }

    /// Encrypt a content stream for a specific object.
    /// The per-object key is derived from the encryption key, object number, and generation number.
    pub fn encryptStream(self: *const SecurityHandler, allocator: Allocator, data: []const u8, obj_num: u32, gen_num: u16) ![]u8 {
        // Compute per-object key: MD5(encryption_key + obj_num_bytes + gen_num_bytes)
        var key_input: [32 + 5]u8 = undefined;
        @memcpy(key_input[0..self.key_length], self.encryption_key[0..self.key_length]);

        // Object number as 3 little-endian bytes
        key_input[self.key_length] = @intCast(obj_num & 0xFF);
        key_input[self.key_length + 1] = @intCast((obj_num >> 8) & 0xFF);
        key_input[self.key_length + 2] = @intCast((obj_num >> 16) & 0xFF);

        // Generation number as 2 little-endian bytes
        key_input[self.key_length + 3] = @intCast(gen_num & 0xFF);
        key_input[self.key_length + 4] = @intCast((gen_num >> 8) & 0xFF);

        const obj_key_len = @min(self.key_length + 5, 16);
        const obj_key_hash = md5_mod.md5(key_input[0 .. self.key_length + 5]);
        const obj_key = obj_key_hash[0..obj_key_len];

        switch (self.algorithm) {
            .rc4_40, .rc4_128 => {
                const output = try allocator.alloc(u8, data.len);
                rc4_mod.rc4(obj_key, data, output);
                return output;
            },
            .aes_128 => {
                // Generate random IV
                var iv: [16]u8 = undefined;
                std.crypto.random.bytes(&iv);
                const encrypted = try aes_mod.aesEncryptCbc(allocator, obj_key, iv, data);

                // Prepend IV to output
                const result = try allocator.alloc(u8, 16 + encrypted.len);
                @memcpy(result[0..16], &iv);
                @memcpy(result[16..], encrypted);
                allocator.free(encrypted);
                return result;
            },
            .aes_256 => {
                var iv: [16]u8 = undefined;
                std.crypto.random.bytes(&iv);
                const encrypted = try aes_mod.aesEncryptCbc(allocator, self.encryption_key[0..32], iv, data);

                const result = try allocator.alloc(u8, 16 + encrypted.len);
                @memcpy(result[0..16], &iv);
                @memcpy(result[16..], encrypted);
                allocator.free(encrypted);
                return result;
            },
        }
    }

    /// Build the /Encrypt dictionary as a PdfObject.
    pub fn buildEncryptDict(self: *const SecurityHandler, allocator: Allocator) !PdfObject {
        var dict = types.pdfDict(allocator);
        errdefer dict.deinit(allocator);

        try dict.dict_obj.put(allocator, "Filter", types.pdfName("Standard"));
        try dict.dict_obj.put(allocator, "V", types.pdfInt(self.algorithm.vValue()));
        try dict.dict_obj.put(allocator, "R", types.pdfInt(self.algorithm.rValue()));
        try dict.dict_obj.put(allocator, "Length", types.pdfInt(@as(i64, @intCast(self.key_length * 8))));
        try dict.dict_obj.put(allocator, "P", types.pdfInt(@as(i64, self.permissions.toInt())));
        try dict.dict_obj.put(allocator, "O", types.pdfString(&self.owner_password_hash));
        try dict.dict_obj.put(allocator, "U", types.pdfString(&self.user_password_hash));

        if (self.algorithm == .aes_128 or self.algorithm == .aes_256) {
            var cf_dict = types.pdfDict(allocator);
            var std_cf = types.pdfDict(allocator);
            try std_cf.dict_obj.put(allocator, "CFM", types.pdfName("AESV2"));
            try std_cf.dict_obj.put(allocator, "AuthEvent", types.pdfName("DocOpen"));
            try std_cf.dict_obj.put(allocator, "Length", types.pdfInt(@as(i64, @intCast(self.key_length))));
            try cf_dict.dict_obj.put(allocator, "StdCF", std_cf);
            try dict.dict_obj.put(allocator, "CF", cf_dict);
            try dict.dict_obj.put(allocator, "StmF", types.pdfName("StdCF"));
            try dict.dict_obj.put(allocator, "StrF", types.pdfName("StdCF"));
        }

        return dict;
    }
};

// -- Tests --

test "security handler: init with defaults" {
    const handler = SecurityHandler.init(.{});
    try std.testing.expect(handler.key_length == 16);
    try std.testing.expect(handler.algorithm == .aes_128);
}

test "security handler: rc4_40" {
    const handler = SecurityHandler.init(.{
        .algorithm = .rc4_40,
        .owner_password = "owner",
        .user_password = "user",
    });
    try std.testing.expect(handler.key_length == 5);
}

test "security handler: permissions to int" {
    const perms = Permissions{};
    const p = perms.toInt();
    // With all permissions set, specific bits should be on
    try std.testing.expect(p & (1 << 2) != 0); // printing
    try std.testing.expect(p & (1 << 4) != 0); // copying
}

test "security handler: permissions restricted" {
    const perms = Permissions{
        .printing = false,
        .copying = false,
        .modifying = false,
        .annotating = false,
        .filling_forms = false,
        .content_accessibility = false,
        .document_assembly = false,
        .high_quality_printing = false,
    };
    const p = perms.toInt();
    try std.testing.expect(p & (1 << 2) == 0); // no printing
    try std.testing.expect(p & (1 << 4) == 0); // no copying
}

test "security handler: build encrypt dict" {
    const allocator = std.testing.allocator;
    const handler = SecurityHandler.init(.{ .algorithm = .rc4_128 });
    var dict = try handler.buildEncryptDict(allocator);
    defer dict.deinit(allocator);

    try std.testing.expect(dict.isDict());
}

test "security handler: encrypt stream rc4" {
    const allocator = std.testing.allocator;
    const handler = SecurityHandler.init(.{
        .algorithm = .rc4_128,
        .user_password = "test",
    });
    const data = "Hello, PDF!";
    const encrypted = try handler.encryptStream(allocator, data, 1, 0);
    defer allocator.free(encrypted);

    try std.testing.expect(encrypted.len == data.len);
    try std.testing.expect(!std.mem.eql(u8, data, encrypted));
}
